File: reflex_home/reflex_home/linear.py
import matplotlib.pyplot as plt
from matplotlib.patches import Rectangle


# 문제 설정
fabric_width = 2000
pieces = [
    (1000, 2000), (1000, 2000), (1000, 2000),
    (660, 2000), (660, 2000), (660, 2000),
    (500, 1000), (500, 1000), (500, 1000), (500, 2000), (300, 500)
    , (300, 500), (200, 900)
]
num_pieces = len(pieces)

# 공간이 비어 있는지 확인하는 함수
def is_space_free(placements, x, y, width, height, fabric_width):
 
    if x + width > fabric_width:
        return False
    
    
    for px, py, pwidth, pheight in placements:
        if not (x + width <= px or x >= px + pwidth or y + height <= py or y >= py + pheight):
            
            return False 
    return True

def append_if_not_exists(arr, item):
    if item not in arr:
        arr.append(item)

# 간극 채워넣기
def fill_gaps(placements, pieces, fabric_width):
    new_placements = placements[:]
    pieces = sorted(pieces, key=lambda piece: piece[0] * piece[1], reverse=True)
    y_positions = [0]  # 초기 y 위치
    max_height = 0
    for width, height in pieces:
        placed = False
 
        y_positions.sort()  # y_positions를 정렬

        for y in y_positions :
            if placed:
                break
            for x in range(fabric_width): 

                if is_space_free(new_placements, x, y, width, height, fabric_width):
                     
                    new_placements.append((x, y, width, height))
                    append_if_not_exists(y_positions,y + height)
                    if max_height < (y + height):
                        max_height = y + height  
                    placed = True
                    break
 
        if not placed:
            # 배치할 수 없으면 새로운 위치에 배치
            current_max_y = max(y + height for _, y, _, height in new_placements)
            new_placements.append((0, current_max_y, width, height))
            append_if_not_exists(y_positions,current_max_y + height) 

        # 기존 배치된 조각 사이의 빈 공간을 확인하고 조각을 넣을 수 있는지 확인
        for i in range(len(new_placements)):
            for j in range(i + 1, len(new_placements)):
                px1, py1, pw1, ph1 = new_placements[i]
                px2, py2, pw2, ph2 = new_placements[j]

                # 가로 방향으로 빈 공간 확인
                if px1 + pw1 <= px2:
                    if is_space_free(new_placements, px1 + pw1, py1, width, height, fabric_width):
                        new_placements.append((px1 + pw1, py1, width, height))
                        placed = True
                        break

                # 세로 방향으로 빈 공간 확인
                if py1 + ph1 <= py2:
                    if is_space_free(new_placements, px1, py1 + ph1, width, height, fabric_width):
                        new_placements.append((px1, py1 + ph1, width, height))
                        placed = True
                        break

            if placed:
                break

    return new_placements

# 배치 시각화
def visualize_placements(placements, fabric_width):
    fig, ax = plt.subplots()
    current_max_y = 0

    for (x, y, width, height) in placements:
        ax.add_patch(Rectangle((x, y), width, height, edgecolor='black', facecolor='gray', fill=True))
        current_max_y = max(current_max_y, y + height)

    ax.set_xlim(0, fabric_width)
    ax.set_ylim(0, current_max_y)
    plt.gca().set_aspect('equal', adjustable='box')
    plt.show()

# 초기 배치
placements = []
print("Placements before filling gaps:", placements)
# 간극 채워넣기
placements = fill_gaps(placements, pieces, fabric_width)
print("Placements after filling gaps:", placements)

# 배치 시각화
visualize_placements(placements, fabric_width)

File: reflex_home/reflex_home/linear.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;

public class FabricCutting
{
    public int FabricWidth = 2000;
    public List<(int width, int height)> Pieces = new List<(int width, int height)>
    {
        (1000, 2000), (1000, 2000), (1000, 2000),
        (660, 2000), (660, 2000), (660, 2000),
        (500, 1000), (500, 1000), (500, 1000), (500, 2000), (300, 500),
        (300, 500), (200, 900)
    };

    public bool IsSpaceFree(List<(int x, int y, int width, int height)> placements, int x, int y, int width, int height, int fabricWidth)
    {
        if (x + width > fabricWidth)
            return false;

        foreach (var (px, py, pwidth, pheight) in placements)
        {
            if (!(x + width <= px || x >= px + pwidth || y + height <= py || y >= py + pheight))
                return false;
        }
        return true;
    }

    public void AppendIfNotExists(List<int> arr, int item)
    {
        if (!arr.Contains(item))
            arr.Add(item);
    }

    public List<(int x, int y, int width, int height)> FillGaps(List<(int x, int y, int width, int height)> placements, List<(int width, int height)> pieces, int fabricWidth)
    {
        var newPlacements = new List<(int x, int y, int width, int height)>(placements);
        pieces = pieces.OrderByDescending(p => p.width * p.height).ToList();
        var yPositions = new List<int> { 0 };
        int maxHeight = 0;

        foreach (var (width, height) in pieces)
        {
            bool placed = false;
            yPositions.Sort();

            foreach (var y in yPositions)
            {
                if (placed)
                    break;
                for (int x = 0; x < fabricWidth; x++)
                {
                    if (IsSpaceFree(newPlacements, x, y, width, height, fabricWidth))
                    {
                        newPlacements.Add((x, y, width, height));
                        AppendIfNotExists(yPositions, y + height);
                        if (maxHeight < y + height)
                            maxHeight = y + height;
                        placed = true;
                        break;
                    }
                }
            }

            if (!placed)
            {
                int currentMaxY = newPlacements.Max(p => p.y + p.height);
                newPlacements.Add((0, currentMaxY, width, height));
                AppendIfNotExists(yPositions, currentMaxY + height);
            }

            for (int i = 0; i < newPlacements.Count; i++)
            {
                for (int j = i + 1; j < newPlacements.Count; j++)
                {
                    var (px1, py1, pw1, ph1) = newPlacements[i];
                    var (px2, py2, pw2, ph2) = newPlacements[j];

                    if (px1 + pw1 <= px2 && IsSpaceFree(newPlacements, px1 + pw1, py1, width, height, fabricWidth))
                    {
                        newPlacements.Add((px1 + pw1, py1, width, height));
                        placed = true;
                        break;
                    }

                    if (py1 + ph1 <= py2 && IsSpaceFree(newPlacements, px1, py1 + ph1, width, height, fabricWidth))
                    {
                        newPlacements.Add((px1, py1 + ph1, width, height));
                        placed = true;
                        break;
                    }
                }
                if (placed)
                    break;
            }
        }

        return newPlacements;
    }

    public void VisualizePlacements(List<(int x, int y, int width, int height)> placements, int fabricWidth)
    {
        Application.Run(new VisualizerForm(placements, fabricWidth));
    }

    public class VisualizerForm : Form
    {
        private List<(int x, int y, int width, int height)> placements;
        private int fabricWidth;

        public VisualizerForm(List<(int x, int y, int width, int height)> placements, int fabricWidth)
        {
            this.placements = placements;
            this.fabricWidth = fabricWidth;
            this.Paint += new PaintEventHandler(DrawPlacements);
            this.Width = fabricWidth / 5;
            this.Height = placements.Max(p => p.y + p.height) / 5 + 50;
        }

        private void DrawPlacements(object sender, PaintEventArgs e)
        {
            Graphics g = e.Graphics;
            foreach (var (x, y, width, height) in placements)
            {
                g.FillRectangle(Brushes.Gray, x / 5, y / 5, width / 5, height / 5);
                g.DrawRectangle(Pens.Black, x / 5, y / 5, width / 5, height / 5);
            }
        }
    }

    public static void Main()
    {
        var fabricCutting = new FabricCutting();
        var placements = new List<(int x, int y, int width, int height)>();
        Console.WriteLine("Placements before filling gaps:");
        Console.WriteLine(string.Join(", ", placements.Select(p => $"({p.x}, {p.y}, {p.width}, {p.height})")));
        
        placements = fabricCutting.FillGaps(placements, fabricCutting.Pieces, fabricCutting.FabricWidth);
        Console.WriteLine("Placements after filling gaps:");
        Console.WriteLine(string.Join(", ", placements.Select(p => $"({p.x}, {p.y}, {p.width}, {p.height})")));

        fabricCutting.VisualizePlacements(placements, fabricCutting.FabricWidth);
    }
}